thunder/core/proxies.py

```diff
@@ -684,7 +699,28 @@ def _elementwise_binary_helper(a, b, name, fn):
             tensor_fn = resolve_method(name, a, b)
             return tensor_fn(a, b)
 
-        return fn(vala, valb)
+        try:
+            method = resolve_method(name, a, b)
+        except Exception as e:
+            return fn(vala, valb)
+
+        if method is None:
+            return fn(vala, valb)
+
+        # circular import
+        from thunder.core.utils import elementwise_type_promotion, dtype_to_numbertype, are_same_dtypes
```

Comment:
Do we expect the type promotion logic here to be deleted in the future? What will that require?

Reply:
We do need the type promotion logic. Maybe not explicitly in the trace if that's what you meant?

My naive understanding is that since we are using `prims` for NumberProxy, it's natural to follow the same protocol that type promotion is handling out side of `prim`. Otherwise, we'll need some special code path inside `prim`'s meta function for NumberProxy. That felts a bit messy to me.

Reply:
Maybe we should be calling `clang` operators that do the type promotion?

Reply:
I think you are saying that `resolve_method(...)` should return `clang.xxx` instead of `prim.xxx` so we don't have to duplicate the type promotion logic here?

Let me dig a bit to see how easy that is :bow:

Reply:
I think so, right? If we want to make these operations on numbers more consistent with operations on tensors.

One thing I'm curious about is what happens in the prim meta operations. Don't those operations have a check that we don't call any symbols from them, but we do a lot of manipulations on numbers? Do we need to disable recording symbols in those operations maybe, and/or check that we don't perform tensor operations?

Reply:
Looks like the check is that the operation is just not on a tensor proxy, so I think we'll be OK with NumberProxies:

https://github.com/Lightning-AI/lightning-thunder/blob/73ca5dc79179773c542d71b3f81d02ae0a220329/thunder/core/prims.py#L41

Reply:
But follow-up question: do operations on NumberProxies called from primitive meta functions get recorded somewhere?

Reply:
Like in the meta operation for reshape:

https://github.com/Lightning-AI/lightning-thunder/blob/73ca5dc79179773c542d71b3f81d02ae0a220329/thunder/core/prims.py#L2976

if the elements of `shape` are NumberProxies, will the multiplications used to compute `numel` be recorded? Is that good or bad?

Reply:
Currently all operations that occur in primitive meta functions are ignored:

https://github.com/Lightning-AI/lightning-thunder/blob/73ca5dc79179773c542d71b3f81d02ae0a220329/thunder/core/symbol.py#L252

Reply:
> if the elements of shape are NumberProxies, will the multiplications used to compute numel be recorded? Is that good or bad?
> Currently all operations that occur in primitive meta functions are ignored:

That does sound like the right behavior in my naive mind. `prims` are the primitive in our trace and we shouldn't further trace anything inside.
In your example, `prims.reshape` does shape manipulation just as a sanity check and that shouldn't be exposed. if we do want things to show up in trace, that should be put inside `clang/__init__.py` I think?
But `prim`'s behavior on NumberProxy needs to be updated... i.e. NumberProxies are going to be dynamic and checks on Meta functions are no longer meaningful. I think it just means certain prim that does a hard check on NumberProxies shouldn't be accepting number proxies any more and we should move those checks to `clang/__init__.py` instead?


Interestingly `clang.reshape` inserts semantic check for cases with `-1` in new shape but not with concrete shapes. I guess it's just assuming `prim` is handling the shape check.
We need to start making conscious decision on what "check" to be baked into the trace.